File: tools/python/category/omim_parsers.py
import re
from typing import Optional, Tuple, Dict, List

LANGUAGES = [
    "af", "ar", "be", "bg", "ca", "cs", "da", "de", "el", "en", "en-GB", "es", "es-MX", "et",
    "eu", "fa", "fi", "fr", "fr-CA", "he", "hi", "hu", "id", "it", "ja", "ko", "lt", "mr", "nb",
    "nl", "pl", "pt", "pt-BR", "ro", "ru", "sk", "sv", "sw", "th", "tr", "uk", "vi", "zh-Hans", "zh-Hant"
]


class AbstractParser:
    def __init__(self, keys: List[str]):
        self.keys = keys

    def parse_line(self, line: str) -> Optional[Tuple[str, str]]:
        raise NotImplementedError("You must implement parse_line.")

    def match_category(self, line: str, result: Dict[str, Dict]):
        category_match = self.category().search(line)
        if category_match:
            category = category_match.group(1)
            if category in self.keys:
                if category not in result:
                    result[category] = {}

    def parse_file(self, filename: str) -> Dict[str, Dict]:
        result = {}
        current_category = None

        with open(filename, "r", encoding="utf-8") as file:
            for line in file:
                line = line.strip()

                # Skip comments and empty lines
                if self.should_exclude_line(line):
                    continue

                # Match a new category
                category_match = self.category().match(line)
                if category_match:
                    current_category = category_match.group(1)
                    if current_category not in result:
                        result[current_category] = {}
                    continue

                # Parse translations for the current category
                if current_category:
                    parsed = self.parse_line(line)
                    if parsed:
                        lang, translation = parsed
                        result[current_category].setdefault(lang, []).append(translation)

        return result

    def category(self) -> re.Pattern:
        raise NotImplementedError("You must implement category.")

    def is_new_category(self, line: str) -> bool:
        return bool(self.category().match(line))

    def extract_category(self, line: str) -> Optional[str]:
        match = self.category().match(line)
        return match.group(1) if match else None

    def should_exclude_line(self, line: str) -> bool:
        return False


class CategoriesParser(AbstractParser):
    def parse_line(self, line: str) -> Optional[Tuple[str, str]]:
        line_match = re.match(r"^([^:]+):(.+)$", line)
        if line_match:
            lang = line_match.group(1).strip()
            translation = line_match.group(2).strip()
            return lang, translation
        return None

    def category(self) -> re.Pattern:
        return re.compile(r"^([a-zA-Z0-9_-]+)\|@(.+)$")

    def should_exclude_line(self, line: str) -> bool:
        return line.startswith("#") or not line


class StringsParser(AbstractParser):
    def parse_line(self, line: str) -> Optional[Tuple[str, str]]:
        line_match = re.match(r"^([^=]+)=(.*)$", line)
        if line_match:
            lang = line_match.group(1).strip()
            translation = line_match.group(2).strip()
            return lang, translation
        return None

    def category(self) -> re.Pattern:
        return re.compile(r"^\[([a-zA-Z0-9_]+)]$")

    def should_exclude_line(self, line: str) -> bool:
        return line.startswith("tags") or not line

File: tools/category_consistency/omim_parsers.rb
module OmimParsers

  # To update the list, run in root directory:
  # sed -nEe "s/ +([a-zA-Z]{2}(-[a-zA-Z]{2,})?) = .*$/\1/p" "data/strings/strings.txt" | sort -u | tr '\n' ' ' | sed -e 's/,$//' | fold -s -w48; echo
  LANGUAGES = %w(af ar be bg ca cs da de el en en-GB es es-MX et
                 eu fa fi fr fr-CA he hi hu id it ja ko lt mr nb
                 nl pl pt pt-BR ro ru sk sv sw th tr uk vi
                 zh-Hans zh-Hant)

  class AbstractParser
    def initialize(keys)
      @keys = keys
    end

    def parse_line(line)
      raise NotImplementedError.new("You must implement parse_file.")
    end

    def match_category(line, result)
      category_match = category.match(line)
      if !category_match.nil?
        category = category_match[1]
        if @keys.include? category
          result[category] ||= {}
        end
      end
    end

    def parse_file(filename)
      current_string = nil
      result = {}
      File.open(filename, 'r:UTF-8').each do |line|
        line.strip!
        next if should_exclude_line? line

        # If line is empty -> next category block started
        if line.empty?
          current_string = nil
          next
        end

        current_string ||= match_category(line, result)

        parsed = parse_line(line)
        if !parsed.nil? and !current_string.nil?
          lang, translation = parsed
          current_string[lang] = translation
        end
      end
      result
    end

    def category
      raise NotImplementedError.new("You must implement category.")
    end

    def should_exclude_line?(line)
      false
    end
  end

  class CategoriesParser < AbstractParser
    def parse_line(line)
      line_match = /^([^:]+):(\S+)$/u.match(line)
      return if !line_match

      lang = $1.strip
      return if !LANGUAGES.include? lang

      translation = $2.strip
      synonyms = []
      translation.split('|').each do |token|
        token_match  = /\d?\^?(.*)$/.match(token)
        synonyms.push(token_match[1]) if token_match
      end
      [lang, synonyms]
    end

    def should_exclude_line?(line)
      line.start_with? '#'
    end

    def category
      # We match only global categories ('food', 'bank'...)
      /^@([A-Za-z0-9]+)$/
    end
  end

  class StringsParser < AbstractParser
    def parse_line(line)
      line_match = /^([^=]+)=(.*)$/.match(line)
      if line_match
        lang = $1.strip
        if LANGUAGES.include? lang
          [lang, $2.strip]
        end
      end
    end

    def category
      /^\[(.+)\]/
    end
  end
end
